pex/pex_builder.py

```diff
@@ -59,6 +61,64 @@ class Value(Enum.Value):
     SYMLINK = Value("symlink")
 
 
+class InvalidZipAppError(Exception):
+    pass
+
+
+class Check(Enum["Check.Value"]):
+    class Value(Enum.Value):
+        def perform_check(
+            self,
+            layout,  # type: Layout.Value
+            path,  # type: str
+        ):
+            # type: (...) -> Optional[bool]
+
+            if self is Check.NONE:
+                return None
+
+            if layout is not Layout.ZIPAPP:
+                return None
+
+            try:
+                importer = zipimport.zipimporter(path)
+
+                # N.B.: The legacy `find_module` method returns the `zipimporter` instance itself on
+                # success and the `find_spec` method returns a `ModuleSpec` instance on success, but
+                # both return `None` on failure to find the module.
+                finder = "find_spec" if hasattr(importer, "find_spec") else "find_module"
+                if getattr(importer, finder)("__main__") is not None:
```

Comment:
Do you have a sense of the performance of this? I assume it won't be running any code within the `__main__` module, and thus will be some O(1) file manipulation/look-ups?

Reply:
Yeah, I actually didn't proceed with this PR until I had done that homework. It's ~100ms for the real case @peachanG hit in Pants (https://github.com/pantsbuild/pants/issues/19505) as documented here: https://github.com/pantsbuild/pex/issues/2247#issuecomment-1741506222

Reply:
Ok, yeah - worse than several seconds by a few orders of magnitude abd the check is also quicker than 100ms by a few order of magnitude (my vetting in the issue did not subtract Python start-up / import time of stdlib stuff which is ~all of the 100ms!). Re-running @peachanG's example at https://github.com/peachanG/pants-sample/tree/whisper with warm pantsd and warm pex cache (already ran `pants package ::` once):
```
$ pants package src/python/main/whisper:main --no-local-cache --pex-verbosity=1
...
pex:     Resolving requirements from lock file thirdparty/python/python-default.lock: 3334.0ms
pex:       Parsing requirements: 7.2ms
pex:       Resolving urls to fetch for 13 requirements from lock thirdparty/python/python-default.lock: 12.7ms
pex:       Downloading 41 distributions to satisfy 13 requirements: 27.5ms
pex:       Categorizing 41 downloaded artifacts: 0.3ms
pex:       Building 3 artifacts and installing 41: 3284.2ms
pex:         Building distributions for: ... 2.8ms
pex:         Calculating project names for direct requirements: ... 0.1ms
pex:         Installing 41 distributions: 3103.6ms
Saving PEX file to src.python.main.whisper/main.pex
pex: Zipping PEX file.
pex: Zipping PEX file.: 305577.9ms

09:10:46.86 [INFO] Wrote dist/src.python.main.whisper/main.pex
```

Now this case is the too large case. In the too many entries case you can imagine a smallish zip64 zip made up of >~65k files that zips relatively quickly. Adding some timing instrumentation to the IT added in this PR:
```
$ tox -qq -epy311 -- --devpi -sk 'test_check and not test_check_resolve_prerelease_transitive_dependencies_issue_1730' --quiet -lwarn
WARNING Logging configured at level WARN.
>>> Create too big file took 3.888437s
>>> Create too big zipapp took 5.139405s
>>> Check warn too-big.pyz.bust-cache took 0.000181s
>>> Check error too-big.pyz.bust-cache took 0.000035s
/home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py311/bin/python: can't find '__main__' module in '/tmp/pytest-of-jsirois/pytest-10/test_check0/too-big.pyz.bust-cache'
>>> Create too many entries zipapp took 0.775023s
>>> Check warn too-many-entries.pyz.bust-cache took 0.000094s
>>> Check error too-many-entries.pyz.bust-cache took 0.000029s
/home/jsirois/dev/pantsbuild/jsirois-pex/.tox/py311/bin/python: can't find '__main__' module in '/tmp/pytest-of-jsirois/pytest-10/test_check0/too-many-entries.pyz.bust-cache'
>>> Create minimal zipapp took 0.000150s
>>> Check OK layout=zipapp check=none took 0.000002s
>>> Check OK layout=zipapp check=warn took 0.000088s
>>> Check OK layout=zipapp check=error took 0.000074s
>>> Check OK layout=packed check=none took 0.000001s
>>> Check OK layout=packed check=warn took 0.000001s
>>> Check OK layout=packed check=error took 0.000001s
>>> Check OK layout=loose check=none took 0.000001s
>>> Check OK layout=loose check=warn took 0.000001s
>>> Check OK layout=loose check=error took 0.000001s
.
1 passed, 685 deselected in 11.40s
```

These represent worse cases and you get:
+ single too big file + __main__.py with no Pex in the way building: ~5s build, <200us check
+ ~65k small python files + __main__.py with no Pex in the way building: ~1s build, <100us to check